pytests/helpers/timeline.md

```diff
@@ -0,0 +1,313 @@
+```py
+initialize = debug_session.send_request('initialize', {'adapterID': 'test'})
+initialize_response = initialize_request.wait_for_response()
+assert (
+    initialize
+    >>
+    Event('initialized', {})
+    >>
+    initialize_response
+) in debug_session.timeline
+```
+Another useful pattern is `>>` combined with `wait_until`:
+```
+initialize = debug_session.send_request('initialize', {'adapterID': 'test'})
+initialized = debug_session.wait_until(Event('initialized'))
+assert (
+    initialize
+    >>
+    Event('output', ANY.dict_with({'category': 'telemetry'}))
+    >>
+    initialized
+) in debug_session.timeline
+```
+
+### Conjuction (`&`)
+
+When two expectations are conjuncted: `(A & B)` - the resulting expectation is realized at the occurrence at which `A` and `B` are both realized, regardless of their relative order. Thus:
+```py
+Event('stopped', ANY) & Event('continued', ANY)
+```
+this expectation will be realized in timeline:
+```py
+('Event', 'continued', {'threadId': 1})
+('Event', 'thread', {'reason': 'exited', 'threadId': 1})
+('Event', 'stopped', {'reason': 'breakpoint', 'threadId': 2})
+```
+but also in a differently ordered timeline:
+```py
+('Event', 'stopped', {'reason': 'breakpoint', 'threadId': 2})
+('Event', 'continued', {'threadId': 1})
+('Event', 'thread', {'reason': 'exited', 'threadId': 1})
+```
+This is most commonly used as seen above, with related events where their relative ordering is unspecified (usually also framed by `>>` to narrow it down to a specific request/response pair). It can also be used to concurrently send multiple requests, and wait until they all got their responses:
+```py
+pause1 = debug_session.send_request('pause', {'threadId': '1'})
+pause2 = debug_session.send_request('pause', {'threadId': '2'})
+debug_session.wait_until(Response(pause1) & Response(pause2))
+```
+
+### Disjunction (`|`)
+
+When two expectations are disjuncted: `(A | B)` - the resulting expectation is realized at the first occurrence at which either `A` or `B` is realized, or both are realized. Thus, the expectation:
+```py
+Event('stopped', ANY) | Event('continued', ANY)
+```
+will be realized in any of these timelines:
+```py
+('Event', 'continued', {'threadId': 1})
+('Event', 'thread', {'reason': 'exited', 'threadId': 1})
+
+('Event', 'stopped', {'reason': 'breakpoint', 'threadId': 2})
+
+('Event', 'continued', {'threadId': 1})
+('Event', 'thread', {'reason': 'exited', 'threadId': 1})
+('Event', 'stopped', {'reason': 'breakpoint', 'threadId': 2})
+
+('Event', 'thread', {'reason': 'exited', 'threadId': 1})
+('Event', 'stopped', {'reason': 'breakpoint', 'threadId': 2})
+('Event', 'continued', {'threadId': 1})
+```
+
+It is usually used to concurrently send multiple requests, and wait until the first one gets a response:
+```py
+pause1 = debug_session.send_request('pause', {'threadId': '1'})
+pause2 = debug_session.send_request('pause', {'threadId': '2'})
+pause_response = debug_session.wait_until(pause1 | pause2)
```

Comment:
>pause1 | pause2 [](start = 42, length = 15)

Should this be Response(pause1) | Resposne(pause2)?

Reply:
Indeed. I'm going to sprinkle asserts somewhat more liberally in the library, since clearly this might be a fairly common mistake, if I'm the one doing it after writing the code :)